lib/srv/git/audit.go

```diff
@@ -0,0 +1,142 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package git
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"log/slog"
+	"sync"
+
+	"github.com/go-git/go-git/v5/plumbing/format/pktline"
+	"github.com/go-git/go-git/v5/plumbing/protocol/packp"
+	"github.com/go-git/go-git/v5/plumbing/transport"
+
+	"github.com/gravitational/teleport"
+	apievents "github.com/gravitational/teleport/api/types/events"
+	"github.com/gravitational/teleport/lib/utils/log"
+)
+
+// CommandRecorder records Git commands by implementing io.Writer to receive a
+// copy of stdin from the git client.
+type CommandRecorder interface {
+	// Writer is the basic interface for the recorder to receive payload.
+	io.Writer
+
+	// GetCommand returns basic info of the command.
+	GetCommand() Command
+	// GetActions returns the action details of the command.
+	GetActions() []*apievents.GitCommandAction
```

Comment:
Those actions are going to be used as audit events, right? If yes, can we identify that there was an error while generating the actions instead of returning empty?

I think that in case of an error, we still want to produce the audit events but with some error message. Only having it return empty seems insufficient in this case, as the caller won't be able to differentiate when the command has no action or if there was an error.

Reply:
updated